File: src/tokenizer.rs
use std::collections::HashMap;

pub struct Tokenizer {}

impl Tokenizer {

    pub fn text_to_sentences( text: &str ) -> Vec<&str> {
        let sentences: Vec<&str> = text.split( "." ).collect() ; 
        sentences
    }

    pub fn sentence_to_tokens( sentence: &str ) -> Vec<&str> {
        let tokens: Vec<&str> = sentence.split_ascii_whitespace().collect() ; 
        tokens
    }

    pub fn get_freq_map<'a>( words: &'a Vec<&'a str> ) -> HashMap<&'a str,usize> {
        let mut freq_map: HashMap<&str,usize> = HashMap::new() ; 
        for word in words.iter() {
            if freq_map.contains_key( word ) {
                freq_map
                    .entry( word )
                    .and_modify( | e | { 
                        *e = *e + 1 
                    } ) ; 
            }
            else {
                freq_map.insert( word.clone() , 1 ) ; 
            }
        }
        freq_map
    }

    pub fn get_weighed_freq_map( freq_map: HashMap<&str,usize> ) -> HashMap<&str,f32> {
        let mut cum_freq: f32 = 0.0 ; 
        freq_map
            .values()
            .into_iter()
            .for_each( | freq | { 
                cum_freq += *freq as f32 ; 
            } ) ; 
        
        let weighed_freq_map: HashMap<&str,f32> = freq_map
            .into_iter()
            .map( | ( k , v ) |  {
                ( k , (v as f32) / cum_freq )
            }
            ).collect() ; 
        weighed_freq_map
    }

}

File: src/main/java/io/shubham0204/text2summary/Tokenizer.kt
/*
 * Copyright (C) 2025 Shubham Panchal
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package io.shubham0204.text2summary

import java.util.regex.Pattern
import kotlin.collections.ArrayList
import kotlin.collections.HashMap

// Helper class which transforms texts to tokens, builds ( word , frequency )  HashMaps.
class Tokenizer {
    companion object {
        // Returns a list of sentences from the given paragraph.
        fun textToSentences(para: String): Array<String> {
            val text = para.trim()
            val pattern =
                Pattern.compile(
                    "[^.!?\\s][^.!?]*(?:[.!?](?!['\"]?\\s|$)[^.!?]*)*[.!?]?['\"]?(?=\\s|$)",
                    Pattern.MULTILINE or Pattern.COMMENTS,
                )
            val matcher = pattern.matcher(text)
            val sentences = ArrayList<String>()
            while (matcher.find()) {
                sentences.add(matcher.group())
            }
            return sentences.toTypedArray()
        }

        // Returns a list of tokens ( words ) from a given sentence ( seq of words ).
        fun sentenceToTokens(s: String): Array<String> {
            val sentence = s.trim().toLowerCase()
            var tokens = sentence.split(" ")
            tokens = tokens.map { Regex("[^A-Za-z0-9 ]").replace(it, "") }
            tokens = tokens.filter { !ENGLISH_STOP_WORDS.contains(it.trim()) }
            tokens = tokens.filter { it.trim().isNotEmpty() and it.trim().isNotBlank() }
            return tokens.toTypedArray()
        }

        // Builds a ( word , frequency ) HashMap.
        fun buildVocab(words: Array<String>): Map<String, Int> {
            val sortedWords = words.toSet()
            val vocab = HashMap<String, Int>()
            for (word in sortedWords) {
                vocab[word] = words.count { it.equals(word) }
            }
            return vocab
        }

        // Builds a ( word , weighted_frequency ) HashMap.
        fun getWeightedVocab(vocab: Map<String, Int>): Map<String, Float> {
            val maxFreq = vocab.values.max().toFloat()
            val weightedFreqHashMap = HashMap<String, Float>()
            vocab.entries.forEach {
                weightedFreqHashMap.put(it.key, it.value.toFloat() / maxFreq)
            }
            return weightedFreqHashMap
        }

        // Removes \n and \r from the given String.
        fun removeLineBreaks(para: String): String =
            para
                .replace("\n", " ")
                .replace("\r", " ")

        // Checks if the compression rate lie in ( 0 , 1 ].
        fun checkRate(rate: Float): Boolean = rate > 0.0 && rate < 1.0

        // Get the indices of top N maximum values in X.
        fun getTopNIndices(
            x: Array<Float>,
            xSorted: Array<Float>,
            N: Int,
        ): Array<Int> {
            val topN = xSorted.take(N)
            val topNIndices = ArrayList<Int>()
            for (i in topN) {
                topNIndices.add(x.indexOf(i))
            }
            topNIndices.sort()
            return topNIndices.distinct().toTypedArray()
        }
    }
}
